### DISCUSSION TITLE
Docblock indent wrongly recognized as 1 space

### DISCUSSION THREAD

**Discussion by Abdillah** [helix-editor/helix#7381] [Troubleshooting]
Helix use automatically detected indent throughout the session due to this change https://github.com/helix-editor/helix/pull/5918. The indentation detection itself is proposed on 2329512122105ab146395b1889b0eed4f2fb522b.
But, helix doesn't correctly recognizing the indent when presented with file containing document block (docblock) comments.

**Package.php**
Edit: Wrong code, please refer to [this comment](https://github.com/helix-editor/helix/discussions/7381#discussioncomment-6225644).

<details>
<summary>The wrong code</summary>

```
<?php
namespace A;

class Package
{
    /**
     * Testing docblock
     *
     * Testing long docblock. Testing long docblock. Testing long docblock. 
     * Testing long docblock. Testing long docblock. Testing long docblock. 
     * Testing long docblock. Testing long docblock. Testing long docblock. 
     * Testing long docblock. Testing long docblock. Testing long docblock. 
     * Testing long docblock. Testing long docblock. Testing long docblock. 
     * Testing long docblock. Testing long docblock. Testing long docblock. 
     * Testing long docblock. Testing long docblock. Testing long docblock. 
     * Testing long docblock. Testing long docblock. Testing long docblock. 
     * Testing long docblock. Testing long docblock. Testing long docblock. 
     * Testing long docblock. Testing long docblock. Testing long docblock. 
     */
    public function __construct()
    {
        // no code
    }
}
```

</details>

This file wrongly recognized as having 1 space indent. While another is fine.
I have seen that Helix use histogram to detect the indent.
Is there anything we can do on this algorithm?

**Comment by pascalkuthe** [comment]
this file is being correctly recognized to have a four space indent  for me

**Reply by Abdillah** [reply]
Oh, it seems so, I think I'm modifying too much (from my private company code).
The most commonly affected is interface with heavy docblock, such as this one.

```
<?php

namespace Some\Namespace\Package;

interface Package
{
    /**
     * Return Product ID catalog that map the package and the ID
     *
     * This is act like variable, but that can't be the case for PHP Enum.
     * Instead, we require a function that returns the map.
     */
    public static function PRODUCT_IDS();

    /**
     * Returns flat array of integers (the ids)
     *
     * Should parse PRODUCT_IDS() into flat array.
     */
    public static function allProductIds(): array;

    /**
     * Given the package, return the ID
     *
     * Some package require additional context or data.
     * If that's the case, calling this function without the data
     * should return the list of possible ids.
     */
    public function productId(): int|array;
}
```

**Reply by pascalkuthe** [reply]
ah now this is reproducible. Indeed the histogram detects changes in the indentation and then uses the most common type of indent. In this case most of the file is madeup of block comments that are indented with a space of 1. There have been other cases where block comments cause issues. The general issue seems to me cases where other languages make up the majority of a file (we treat comments as an injected language) I think the right soulution would be to ignore texet inside injections if a TS grammar is available in the histogram. To make sure we get good indentation inside the injections we could just always use the language setting inside the injection (auto detection/manual setting won't work there since we have no way to persistently associate data with an injection)

**Reply by Abdillah** [reply]
The "injected language" is the one defined here ([injection.scm](https://github.com/helix-editor/helix/blob/master/runtime/queries/php/injections.scm)) isn't it. For comments to be defined as injected kind of deviate from the notion that it is part of the language. But useful enough if it might support for markup in docblock, etc.
